File: targets/java/micronaut/src/main/java/rb/micronaut/routes/Middleware.java
package rb.micronaut.routes;

import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Get;
import rb.domain.Domain;
import rb.domain.Model.PayloadBody;

/**
 * middleware: per-layer dispatch cost at 0, 4 and 16 no-op layers.
 *
 * Micronaut's layer is an @Filter, and the annotation takes the patterns it applies to,
 * which is the scoping the family needs. Each layer runs and does nothing else.
 *
 * The layers themselves are in filters/: Micronaut resolves them from annotations at
 * compile time, so one bean is one layer and four and sixteen are written out.
 */
@Controller
public class Middleware {

  @Get("/middleware/none")
  PayloadBody none() {
    return Domain.payload("small");
  }

  // rb:snippet middleware.four
  @Get("/middleware/four")
  PayloadBody four() {
    return Domain.payload("small");
  }

  // rb:snippet middleware.sixteen
  @Get("/middleware/sixteen")
  PayloadBody sixteen() {
    return Domain.payload("small");
  }
}

File: targets/go/chi/middleware.go
// middleware: per-layer dispatch cost at 0, 4 and 16 no-op layers.
//
// chi's r.With returns a router carrying those middlewares and nothing else, which is the
// scoping the family needs. Each layer calls the next and does nothing else.
package main

import (
	"net/http"

	"github.com/go-chi/chi/v5"
)

// rb:wiring middleware.*
func noop(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
		next.ServeHTTP(w, req)
	})
}

// rb:wiring middleware.*
func layers(n int) []func(http.Handler) http.Handler {
	out := make([]func(http.Handler) http.Handler, n)
	for i := range out {
		out[i] = noop
	}
	return out
}

func registerMiddleware(r chi.Router) {
	r.Get("/middleware/none", payload("small"))

	r.With(layers(4)...).Get("/middleware/four", payload("small"))

	r.With(layers(16)...).Get("/middleware/sixteen", payload("small"))
}
